Mathlib/Algebra/Order/Archimedean/Class.lean

```diff
@@ -0,0 +1,410 @@
+
+/-!
+# Archimedean classes of a linearly ordered group
+
+This file defines archimedean classes of a given linearly ordered group. Archimedean classes
+measure to what extent the group fails to be Archimedean. For additive group, elements `a` and `b`.
+in the same class are "equivalent" in the sense that there exist two natural numbers
+`m` and `n` such that `|a| ≤ m • |b|` and `|b| ≤ n • |a|`. An element `a` in a higher class than `b`
+is "infinitesimal" to `b` in the sense that `n • |a| < |b|` for all natural number `n`.
+
+## Main definitions
+
+* `ArchimedeanClass` is the archimedean class for additive linearly ordered group.
+* `MulArchimedeanClass` is the archimedean class for multiplicative linearly ordered group.
+* `ArchimedeanClass.orderHom` and `MulArchimedeanClass.orderHom` are `OrderHom` over
+  archimedean classes lifted from ordered group homomorphisms.
+
+## Main statements
+
+The following theorems state that an ordered commutative group is (mul-)archimedean if and only if
+all non-identity elements belong to the same (`Mul`-)`ArchimedeanClass`:
+* `ArchimedeanClass.archimedean_of_mk_eq_mk` / `MulArchimedeanClass.mulArchimedean_of_mk_eq_mk`
+* `ArchimedeanClass.mk_eq_mk_of_archimedean` / `MulArchimedeanClass.mk_eq_mk_of_mulArchimedean`
+
+## Implementation notes
+
+Archimedean classes are equipped with a linear order, where elements with smaller absolute value
+are placed in a *higher* classes by convention. Ordering backwards this way simplifies
+formalization of theorems such as the Hahn embedding theorem.
+
+To naturally derive this order, we first define it on the underlying group via the type
+synonym (`Mul`-)`ArchimedeanOrder`, and define (`Mul`-)`ArchimedeanClass` as `Antisymmetrization` of
+the order.
+
+-/
+
+section ArchimedeanOrder
+variable {M : Type*}
+
+variable (M) in
+/-- Type synonym to equip a ordered group with a new `Preorder` defined by the infinitesimal order
+of elements. `a` is said less than `b` if `b` is infinitesimal comparing to `a`, or more precisely,
+`∀ n, |b|ₘ ^ n < |a|ₘ`. If `a` and `b` are neither infinitesimal to each other, they are equivalent
+in this order. -/
+@[to_additive ArchimedeanOrder
+"Type synonym to equip a ordered group with a new `Preorder` defined by the infinitesimal order
+of elements. `a` is said less than `b` if `b` is infinitesimal comparing to `a`, or more precisely,
+`∀ n, n • |b| < |a|`. If `a` and `b` are neither infinitesimal to each other, they are equivalent
+in this order."]
+def MulArchimedeanOrder := M
+
+namespace MulArchimedeanOrder
+
+/-- Create a `MulArchimedeanOrder` element from the underlying type. -/
+@[to_additive "Create a `ArchimedeanOrder` element from the underlying type."]
+def of (a : M) : MulArchimedeanOrder M := a
+
+/-- Retrieve the underlying value from a `MulArchimedeanOrder` element. -/
+@[to_additive "Retrieve the underlying value from a `ArchimedeanOrder` element."]
+def val (a : MulArchimedeanOrder M) : M := a
```

Comment:
For type synonyms, we usually set up an equiv between the old and new type. See for example https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/TypeTags/Basic.html#Multiplicative.ofAdd